stock_partner_delivery_window/models/res_partner.py

```diff
@@ -180,3 +183,76 @@ def copy_data(self, default=None):
         ]
         result["delivery_time_window_ids"] = [(0, 0, val) for val in values]
         return [result]
+
+    def next_delivery_window_start_datetime(self, from_date=None, timedelta_days=None):
+        """Get next starting datetime in a preferred delivery window.
+
+        If from_date is already in a delivery window, from_date is "the next"
+
+        :param from_date: Datetime object (Leave empty to use now())
+        :param timedelta_days: Number of days to use in the computation
+                               (Leave empty to use 7 days or 1 week)
+        :return: Datetime object
+        """
+        self.ensure_one()
+        if not from_date:
+            from_date = datetime.now()
+        if self.is_in_delivery_window(from_date):
+            return fields.Datetime.to_datetime(from_date)
+        if timedelta_days is None:
+            timedelta_days = 7
+        if self.delivery_time_preference == "workdays":
+            datetime_windows = self.get_next_workdays_datetime(
+                from_date, from_date + timedelta(days=timedelta_days)
+            )
+        else:
+            datetime_windows = self.get_next_windows_start_datetime(
+                from_date, from_date + timedelta(days=timedelta_days)
+            )
+        for dwin_start in datetime_windows:
+            if dwin_start >= from_date:
+                return dwin_start
+        raise UserError(
+            _("Something went wrong trying to find next delivery window. Date: %s")
+            % str(from_date)
+        )
+
+    def get_next_workdays_datetime(self, from_datetime, to_datetime):
+        """Returns all the delivery windows in the provided date range.
+
+        :param from_datetime: Datetime object
+        :param to_datetime: Datetime object
+        :return: List of Datetime objects
+        """
+        dates = date_range(from_datetime, to_datetime, timedelta(days=1))
+        return [date for date in dates if date.weekday() < 5]
+
+    def get_next_windows_start_datetime(self, from_datetime, to_datetime):
+        """Get all delivery windows start time.
+
+        Range from from_datetime weekday to to_datetime weekday.
+
+        Note result can include a start datetime that is before from_datetime
+        on the from_datetime weekday
+
+        :param from_datetime: Datetime object
+        :param to_datetime: Datetime object
+        :return: List of Datetime objects
+        """
+        res = list()
+        for this_datetime in date_range(from_datetime, to_datetime, timedelta(days=1)):
+            this_weekday_number = this_datetime.weekday()
+            this_weekday = self.env["time.weekday"].search(
+                [("name", "=", this_weekday_number)], limit=1
+            )
```

Comment:
When the date_range is big you'll have many many searches.
I'd recommend doing the following:

```suggestion
        dts = date_range(from_datetime, to_datetime, timedelta(days=1))
        weekdays = {x.name: x for x in self.env["time.weekday"].search([("name", "in", dts)])}
        for this_datetime, this_weekday in weekdays.items():
            this_weekday_number = this_datetime.weekday()
```